### PR TITLE
Move moniker related functionality to MonikerProvider

### PR DESCRIPTION
**Pull request by OsmondJiang**
move moniker range parer and compare to moniker provider

### CODE REVIEW

File: src/docfx/build/moniker/MonikersProvider.cs [hunk 1]
```diff
@@ -3,23 +3,44 @@
...
             _rules.Reverse();
+
+            var monikerDefinition = new MonikerDefinitionModel();
+            if (!string.IsNullOrEmpty(config.MonikerDefinition))
+            {
+                var path = docset.GetFileRestorePath(config.MonikerDefinition);
+                monikerDefinition = JsonUtility.Deserialize<MonikerDefinitionModel>(File.ReadAllText(path));
+            }
+            _rangeParser = new MonikerRangeParser(monikerDefinition);
+            _monikerAscendingComparer = new Lazy<MonikerComparer>(() => new MonikerComparer(monikerDefinition));
```
Comment (live1206) [inline src/docfx/build/moniker/MonikersProvider.cs]: `Lazy<>` is not needed here, right? just initiate a new comparer

Comment (live1206) [reply src/docfx/build/moniker/MonikersProvider.cs]: Also, we have `Lazy<MonikersProvider>` in the outer layer

File: src/docfx/build/moniker/MonikersProvider.cs [hunk 2]
```diff
@@ -3,23 +3,44 @@
...
             foreach (var (key, monikerRange) in config.MonikerRange)
             {
                 _rules.Add((GlobUtility.CreateGlobMatcher(key), monikerRange));
             }
             _rules.Reverse();
+
+            var monikerDefinition = new MonikerDefinitionModel();
+            if (!string.IsNullOrEmpty(config.MonikerDefinition))
+            {
+                var path = docset.GetFileRestorePath(config.MonikerDefinition);
```
Comment (live1206) [inline src/docfx/build/moniker/MonikersProvider.cs]: It should be enough to just input path here, avoid exposing the entire Docset

File: src/docfx/build/moniker/MonikersProvider.cs [hunk 3]
```diff
@@ -3,23 +3,44 @@
...
     {
         private readonly List<(Func<string, bool> glob, string monikerRange)> _rules = new List<(Func<string, bool>, string)>();
+        private readonly MonikerRangeParser _rangeParser;
+        private readonly Lazy<MonikerComparer> _monikerAscendingComparer;
+        private readonly Lazy<MonikerComparer> _monikerDescendingComparer;
 
-        public MonikersProvider(Config config)
+        public MonikerComparer AscendingComparer => _monikerAscendingComparer.Value;
+
+        public MonikerComparer DescendingComparer => _monikerDescendingComparer.Value;
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikersProvider.cs]: a single `Comparer` is enough. There is no `StringDescendingComparer` for example.  The only use of `DescendingComparer` is in `XrefMap` which should be using `OrderByDescending` over with a comparer.

File: src/docfx/build/moniker/MonikersProvider.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/build/moniker/MonikersProvider.cs]: Taking a `Docset` as input is fine, `config` can be accessed from `docset` so you just need a `docset` as the input.

File: src/docfx/build/moniker/MonikersProvider.cs [hunk 4]
```diff
@@ -3,23 +3,44 @@
...
+            {
+                var path = docset.GetFileRestorePath(config.MonikerDefinition);
+                monikerDefinition = JsonUtility.Deserialize<MonikerDefinitionModel>(File.ReadAllText(path));
+            }
+            _rangeParser = new MonikerRangeParser(monikerDefinition);
+            _monikerAscendingComparer = new Lazy<MonikerComparer>(() => new MonikerComparer(monikerDefinition));
+            _monikerDescendingComparer = new Lazy<MonikerComparer>(() => new MonikerComparer(monikerDefinition, false));
         }
 
+        public IReadOnlyList<string> GetRangeMonikers(string rangeString)
```
Comment (yufeih) [inline src/docfx/build/moniker/MonikersProvider.cs]: use `_monikerRangeParser.Parse()` inline is clear enough

File: src/docfx/build/moniker/MonikersProvider.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/build/moniker/MonikersProvider.cs]: ok


---
In reply to: [237374624](https://github.com/dotnet/docfx/pull/3695#discussion_r237374624) [](ancestors = 237374624)

**Pull request merged**